protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnConverter.java

```diff
@@ -0,0 +1,122 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.processor.base;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+
+import org.apache.james.mailbox.MessageUid;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Ordering;
+
+public class UidMsnConverter {
+
+    public static final int FIRST_MSN = 1;
+    private static final Logger LOGGER = LoggerFactory.getLogger(UidMsnConverter.class);
+
+    private final ArrayList<MessageUid> uids;
+
+    public UidMsnConverter(Iterator<MessageUid> iterator) {
+        uids = Lists.newArrayList(iterator);
+        Collections.sort(uids);
+    }
+
+    public synchronized Optional<Integer> getMsn(MessageUid uid) {
+        try {
+            int position = Ordering.explicit(uids).binarySearch(uids, uid);
+            return Optional.of(position + 1);
+        } catch (Exception e) {
+            LOGGER.trace("Error while converting uid to msn, uid not found: " + uid.asLong());
+            return Optional.absent();
+        }
+    }
+
+    public synchronized Optional<MessageUid> getUid(int msn) {
+        if (msn <= uids.size() && msn > 0) {
+            return Optional.of(uids.get(msn - 1));
+        }
+        return Optional.absent();
+    }
+
+    public synchronized Optional<MessageUid> getLastUid() {
+        if (uids.isEmpty()) {
+            return Optional.absent();
+        }
+        return getUid(getLastMsn());
+    }
+
+    public synchronized Optional<MessageUid> getFirstUid() {
+        return getUid(FIRST_MSN);
+    }
+
+    public synchronized int getNumMessage() {
+        return uids.size();
+    }
+
+    public synchronized void remove(MessageUid uid) {
+        uids.remove(uid);
+    }
+
+    public synchronized boolean isEmpty() {
+        return uids.isEmpty();
+    }
+
+    public synchronized void clear() {
+        uids.clear();
+    }
+
+    public synchronized void addUid(MessageUid uid) {
+        if (uids.contains(uid)) {
+            return;
+        }
+        if (isLastUid(uid)) {
+            uids.add(uid);
+        } else {
+            uids.add(uid);
+            Collections.sort(uids);
+        }
+    }
+
+    private boolean isLastUid(MessageUid uid) {
+        Optional<MessageUid> lastUid = getLastUid();
+        return !lastUid.isPresent() ||
+            lastUid.get().compareTo(uid) < 0;
+    }
+
+    @VisibleForTesting
+    ImmutableBiMap<Integer, MessageUid> getConversion() {
```

Comment:
I prefer this to say "The internal structure is sorted", which I don't care, from a business perspective.

I want to say:

```
MSN 1 -> UID 2
MSN 2 -> UID 4
MSN 3 -> UID 6
```

And only these one.

Maybe we can also do it with a map. But I would feel asserting the internal data structure, ie:

```
assertThat(uidList).isSorted()
```

Is less pertinent.

Here the test is too bound to the internal of the object, and don't abstract it at all. What if we decide to handle it without a sorted list? If we ever found some better datastructure?